File: javascript/resistor-color-trio/resistor-color-trio.js
// reuse the "resistor color" exercise
import { colorCode } from '../resistor-color/resistor-color.js';

export class ResistorColorTrio {
  constructor(colors) {
    let x, y, z;
    try {
      [x, y, z] = colors.map(c => colorCode(c));
    } catch (e) {
      throw new Error("invalid color");
    }
    this.value = (10 * x + y) * (10 ** z);
  }

  get label() {
    let val = this.value;
    let magnitude = 0;
    while (val > 0 && val % 1000 == 0) {
      val /= 1000;
      magnitude++;
    }
    const prefix = ["", "kilo", "mega", "giga"][magnitude];

    return `Resistor value: ${val} ${prefix}ohms`;
  }
}

File: lua/resistor-color-trio/resistor-color-trio.lua
local rc = require 'resistor-color'
local rcd = require 'resistor-color-duo'

local prefix = { "", "kilo", "mega", "giga" }


function decode(c1, c2, c3)
  local val = rcd.value({c1, c2}) * 10 ^ rc.color_code(c3)
  local idx = 1

  while val > 0 and val % 1000 == 0 do
    val = val / 1000
    idx = idx + 1
  end

  return val, prefix[idx].."ohms"
end


return {
  decode = decode
}
